lookout/style/cloner.py

```diff
@@ -0,0 +1,73 @@
+"""Cloning utilities for git repositories."""
+
+from itertools import repeat
+import logging
+import multiprocessing
+import os
+from typing import Dict, Optional, Sequence, Tuple
+
+from dulwich import porcelain
+
+
+class Cloner:
+    """
+    Class to clone a list of git repositories in parallel.
+    """
+
+    _log = logging.getLogger("Cloner")
+
+    def __init__(self, location: str, processes: Optional[int] = None):
+        """
+        Create new Cloner instance.
+
+        :param location: Path to the dirrectory where repositories is saved.
+        :param processes: Number of processes to use. 'os.cpu_count()' is used by default.
+        """
+        os.makedirs(location, exist_ok=True)
+        self._location = location
+        self._processes = processes
+
+    def clone_repositories(self, repositories: Sequence[str]) -> Dict[str, str]:
+        """
+        Run repositories cloning in parallel.
+
+        :param repositories: List of URLs to clone.
+        :return: Mapping from provided URL to the location in the file system for successfully \
+                 downloaded repositories.
+        """
+        self._log.info("start cloning %d repositories", len(repositories))
+        with multiprocessing.Pool(self._processes) as pool:
+            repo_paths = pool.map(self._clone_repository,
+                                  zip(repositories, repeat(self._location)))
+        repositories_dir = {repo: git_dir for repo, git_dir in repo_paths if git_dir}
+        self._log.info("successfully cloned %d/%d repositories",
+                       len(repositories_dir), len(repositories))
+        return repositories_dir
+
+    @staticmethod
+    def get_repo_name(url: str) -> str:
+        """
+        Convert URL to repository name.
+
+        :param url: Repository URL to get name for.
+        :return: Generated repository name.
+        """
+        return "-".join(url.split("/")[-2:])
```

Comment:
We should rather use `@` so that we can reliably recover the repo name. We don't  need this  now, but still

Reply:
Ok, I am going to include `github.com` to the name too